File: articles/the-history-of-personality-psychology-from-characterology-to-personality-science/r/analyze_personality_history.R
args <- commandArgs(trailingOnly = FALSE)
file_arg <- grep("^--file=", args, value = TRUE)
script_path <- if (length(file_arg) > 0) normalizePath(sub("^--file=", "", file_arg[1])) else normalizePath("r/analyze_personality_history.R")
root <- normalizePath(file.path(dirname(script_path), ".."))
data_path <- file.path(root, "data", "synthetic_personality_history_items.csv")
outputs <- file.path(root, "outputs")
dir.create(outputs, showWarnings = FALSE, recursive = TRUE)

data <- read.csv(data_path, stringsAsFactors = FALSE)

items <- paste0("item", 1:60)
c_items <- paste0("c", 1:6)
traits <- c("extraversion_score", "agreeableness_score", "conscientiousness_score", "neuroticism_score", "openness_score")
history <- c("characterology_typology_index", "psychometric_structure_index", "person_situation_index", "narrative_identity_index", "measurement_invariance_caution_index", "historical_method_maturity_index")

alpha_score <- function(df) {
  k <- ncol(df)
  item_vars <- apply(df, 2, var)
  total_var <- var(rowSums(df))
  (k / (k - 1)) * (1 - sum(item_vars) / total_var)
}

data$conscientiousness_rescored <- rowMeans(data[, c_items])
data$stability_change <- data$conscientiousness_t2 - data$conscientiousness_t1

reliability <- data.frame(
  scale = c("conscientiousness_c1_c6", "full_personality_item_pool_item1_item60"),
  level = c("short_trait_scale", "broad_item_pool"),
  n_items = c(6, 60),
  cronbach_alpha = c(alpha_score(data[, c_items]), alpha_score(data[, items]))
)

# PCA using base R.
scaled_items <- scale(data[, items])
pca <- prcomp(scaled_items, center = FALSE, scale. = FALSE)
component_scores <- as.data.frame(pca$x[, 1:10])
names(component_scores) <- paste0("component_", 1:10)
data <- cbind(data, component_scores)

pca_summary <- data.frame(
  component = 1:10,
  explained_variance_ratio = (pca$sdev[1:10]^2) / sum(pca$sdev^2),
  cumulative_explained_variance = cumsum((pca$sdev[1:10]^2) / sum(pca$sdev^2))
)

data$person_situation_interaction <- data$trait_score * data$situation_strength

models <- list(
  conscientiousness_stability_t2_from_t1 = lm(conscientiousness_t2 ~ conscientiousness_t1, data = data),
  behavior_from_trait_situation_interaction = lm(behavior_score ~ trait_score + situation_strength + person_situation_interaction, data = data),
  psychometric_structure_from_traits = lm(psychometric_structure_index ~ extraversion_score + agreeableness_score + conscientiousness_score + neuroticism_score + openness_score, data = data),
  historical_method_maturity = lm(historical_method_maturity_index ~ psychometric_structure_index + person_situation_index + narrative_identity_index + measurement_invariance_caution_index, data = data),
  typology_residual = lm(characterology_typology_index ~ psychometric_structure_index + historical_method_maturity_index, data = data)
)

model_fit <- data.frame(
  model = names(models),
  r_squared = sapply(models, function(m) summary(m)$r.squared),
  adj_r_squared = sapply(models, function(m) summary(m)$adj.r.squared),
  n = nrow(data)
)

coef_table <- function(model, model_name) {
  coefs <- summary(model)$coefficients
  data.frame(model = model_name, term = rownames(coefs), estimate = coefs[,1], standard_error = coefs[,2], t_value = coefs[,3], p_value = coefs[,4], row.names = NULL)
}

model_coefficients <- do.call(rbind, Map(coef_table, models, names(models)))

stability_summary <- data.frame(
  measure = "conscientiousness_rank_order_stability",
  correlation = cor(data$conscientiousness_t1, data$conscientiousness_t2),
  mean_t1 = mean(data$conscientiousness_t1),
  mean_t2 = mean(data$conscientiousness_t2),
  mean_change = mean(data$stability_change),
  sd_change = sd(data$stability_change)
)

write.csv(summary(data[, c(traits, "conscientiousness_t1", "conscientiousness_t2", "behavior_score", "trait_score", "situation_strength")]), file.path(outputs, "r_trait_summary.csv"))
write.csv(summary(data[, c(history, "stability_change")]), file.path(outputs, "r_history_index_summary.csv"))
write.csv(cor(data[, c(traits, history, "conscientiousness_t1", "conscientiousness_t2", "behavior_score", "trait_score", "situation_strength")], use = "pairwise.complete.obs"), file.path(outputs, "r_personality_history_correlations.csv"))
write.csv(pca_summary, file.path(outputs, "r_pca_dimensionality_summary.csv"), row.names = FALSE)
write.csv(reliability, file.path(outputs, "r_reliability_summary.csv"), row.names = FALSE)
write.csv(model_fit, file.path(outputs, "r_model_fit_summary.csv"), row.names = FALSE)
write.csv(model_coefficients, file.path(outputs, "r_model_coefficients.csv"), row.names = FALSE)
write.csv(stability_summary, file.path(outputs, "r_stability_summary.csv"), row.names = FALSE)
write.csv(data, file.path(outputs, "r_scored_personality_history_items.csv"), row.names = FALSE)
cat("Wrote R outputs to:", outputs, "\n")

File: articles/the-history-of-personality-psychology-from-characterology-to-personality-science/python/analyze_personality_history.py
#!/usr/bin/env python3
from pathlib import Path
import json
import numpy as np
import pandas as pd

ROOT = Path(__file__).resolve().parents[1]
DATA = ROOT / "data" / "synthetic_personality_history_items.csv"
OUT = ROOT / "outputs"
OUT.mkdir(exist_ok=True)

ITEMS = [f"item{i}" for i in range(1, 61)]
C_ITEMS = [f"c{i}" for i in range(1, 7)]
TRAITS = [
    "extraversion_score",
    "agreeableness_score",
    "conscientiousness_score",
    "neuroticism_score",
    "openness_score",
]
HISTORY = [
    "characterology_typology_index",
    "psychometric_structure_index",
    "person_situation_index",
    "narrative_identity_index",
    "measurement_invariance_caution_index",
    "historical_method_maturity_index",
]
LONGITUDINAL = ["conscientiousness_t1", "conscientiousness_t2"]
INTERACTION = ["behavior_score", "trait_score", "situation_strength", "person_situation_interaction"]
REQUIRED = {"participant_id", *ITEMS, *C_ITEMS, *TRAITS, *HISTORY, *LONGITUDINAL, *INTERACTION}

def cronbach_alpha(frame: pd.DataFrame) -> float:
    clean = frame.dropna()
    n_items = clean.shape[1]
    if n_items <= 1:
        return float("nan")
    item_variances = clean.var(axis=0, ddof=1)
    total_score = clean.sum(axis=1)
    total_var = total_score.var(ddof=1)
    return float((n_items / (n_items - 1)) * (1 - item_variances.sum() / total_var)) if total_var else float("nan")

def ols(df: pd.DataFrame, y: str, xs: list[str]) -> dict:
    model_df = df[[y] + xs].dropna()
    X = np.column_stack([np.ones(len(model_df)), model_df[xs].to_numpy(float)])
    beta, *_ = np.linalg.lstsq(X, model_df[y].to_numpy(float), rcond=None)
    pred = X @ beta
    ss_res = float(np.sum((model_df[y].to_numpy(float) - pred) ** 2))
    ss_tot = float(np.sum((model_df[y].to_numpy(float) - model_df[y].mean()) ** 2))
    r2 = 1.0 - ss_res / ss_tot if ss_tot else float("nan")
    return {"coefficients": dict(zip(["intercept", *xs], map(float, beta))), "r_squared": r2, "n": int(len(model_df))}

def standardized_pca_variance(frame: pd.DataFrame, n_components: int = 10) -> tuple[pd.DataFrame, np.ndarray, pd.DataFrame]:
    X = frame.to_numpy(float)
    X = (X - X.mean(axis=0)) / X.std(axis=0, ddof=1)
    cov = np.cov(X, rowvar=False)
    eigvals, eigvecs = np.linalg.eigh(cov)
    order = np.argsort(eigvals)[::-1]
    eigvals = eigvals[order]
    eigvecs = eigvecs[:, order]
    ratios = eigvals / eigvals.sum()
    scores = X @ eigvecs[:, :n_components]
    summary = pd.DataFrame({
        "component": list(range(1, n_components + 1)),
        "explained_variance_ratio": ratios[:n_components],
        "cumulative_explained_variance": np.cumsum(ratios[:n_components]),
    })
    loadings = pd.DataFrame(
        eigvecs[:, :n_components],
        index=frame.columns,
        columns=[f"component_{i}" for i in range(1, n_components + 1)],
    )
    return summary, scores, loadings

df = pd.read_csv(DATA)
missing = REQUIRED - set(df.columns)
if missing:
    raise ValueError(f"Dataset missing columns: {sorted(missing)}")

df["conscientiousness_rescored"] = df[C_ITEMS].mean(axis=1)
df["stability_change"] = df["conscientiousness_t2"] - df["conscientiousness_t1"]

pca_summary, component_scores, loadings = standardized_pca_variance(df[ITEMS], n_components=10)
for i in range(component_scores.shape[1]):
    df[f"component_{i+1}"] = component_scores[:, i]

trait_summary = df[TRAITS + LONGITUDINAL + INTERACTION].agg(["mean", "std", "min", "max"]).T.reset_index().rename(columns={"index": "variable"})
history_summary = df[HISTORY + ["stability_change"]].agg(["mean", "std", "min", "max"]).T.reset_index().rename(columns={"index": "variable"})
correlations = df[TRAITS + HISTORY + LONGITUDINAL + INTERACTION].corr()

reliability = pd.DataFrame([
    {"scale": "conscientiousness_c1_c6", "level": "short_trait_scale", "n_items": 6, "cronbach_alpha": cronbach_alpha(df[C_ITEMS])},
    {"scale": "full_personality_item_pool_item1_item60", "level": "broad_item_pool", "n_items": 60, "cronbach_alpha": cronbach_alpha(df[ITEMS])},
])

models = {
    "conscientiousness_stability_t2_from_t1": ols(df, "conscientiousness_t2", ["conscientiousness_t1"]),
    "behavior_from_trait_situation_interaction": ols(df, "behavior_score", ["trait_score", "situation_strength", "person_situation_interaction"]),
    "psychometric_structure_from_traits": ols(df, "psychometric_structure_index", TRAITS),
    "historical_method_maturity_from_components": ols(df, "historical_method_maturity_index", ["psychometric_structure_index", "person_situation_index", "narrative_identity_index", "measurement_invariance_caution_index"]),
    "typology_residual_from_low_method_maturity": ols(df, "characterology_typology_index", ["psychometric_structure_index", "historical_method_maturity_index"]),
    "professional_use_boundary": "Suitable for education, research prototyping, psychometric demonstration, consulting support, organizational learning, coaching reflection, and methodological comparison; not a standalone assessment, diagnosis, placement, screening, moral-labeling, prediction, or decision system.",
}

model_fit = pd.DataFrame([
    {"model": name, "r_squared": spec["r_squared"], "n": spec["n"]}
    for name, spec in models.items()
    if isinstance(spec, dict)
])

stability_correlation = float(df[["conscientiousness_t1", "conscientiousness_t2"]].corr().iloc[0, 1])
stability_summary = pd.DataFrame([
    {
        "measure": "conscientiousness_rank_order_stability",
        "correlation": stability_correlation,
        "mean_t1": df["conscientiousness_t1"].mean(),
        "mean_t2": df["conscientiousness_t2"].mean(),
        "mean_change": df["stability_change"].mean(),
        "sd_change": df["stability_change"].std(ddof=1),
    }
])

trait_summary.to_csv(OUT / "python_trait_summary.csv", index=False)
history_summary.to_csv(OUT / "python_history_index_summary.csv", index=False)
correlations.to_csv(OUT / "python_personality_history_correlations.csv")
pca_summary.to_csv(OUT / "python_pca_dimensionality_summary.csv", index=False)
loadings.to_csv(OUT / "python_pca_loadings.csv")
reliability.to_csv(OUT / "python_reliability_summary.csv", index=False)
model_fit.to_csv(OUT / "python_model_fit_summary.csv", index=False)
stability_summary.to_csv(OUT / "python_stability_summary.csv", index=False)
(OUT / "python_models.json").write_text(json.dumps(models, indent=2))
df.to_csv(OUT / "python_scored_personality_history_items.csv", index=False)
print(f"Wrote outputs to {OUT}")
